addons/website_sale_coupon/wizard/sale_coupon_share.py

```diff
@@ -0,0 +1,113 @@
+
+    @api.model
+    def try_on_order(self, order, program, coupon=None):
+        self.ensure_one()
+
+        error_status = {}
+        if coupon and coupon not in order.applied_coupon_ids:
+            error_status = coupon.apply_on_order(order)
+        elif program and program != order.code_promo_program_id:
+            error_status = program.apply_on_order(order, self.promo_code)
+
+        return error_status
+
+    @api.constrains('coupon_id', 'program_id')
+    def _check_program(self):
+        if self.filtered(lambda record: not record.coupon_id and record.program_id.program_type == 'coupon_program'):
+            raise ValidationError('A coupon is needed for coupon programs')
+
+    @api.constrains('website_id', 'program_id')
+    def _check_website(self):
+        if self.filtered(lambda record: record.program_website_id and record.program_website_id != record.website_id):
+            raise ValidationError('The shared website should correspond to the website of the program')
+
+    @api.depends('coupon_id.code', 'program_id.promo_code')
+    def _get_promo_code(self):
+        for record in self:
+            record.promo_code = record.coupon_id.code or record.program_id.promo_code
+
+    @api.depends('website_id', 'promo_code', 'create_link_tracker')
+    def _get_share_link(self):
+        for record in self:
+            target_url = '{base}/coupon/{code}?{query}'.format(
+                base=record.website_id.get_base_url(),
+                code=record.promo_code,
+                query=url_encode({'path': record.redirect}),
+            )
+
+            if record.create_link_tracker:
+                tracker = self.env['link.tracker'].search([(
+                    'url', '=', target_url
+                )], limit=1) or self.env['link.tracker'].create({
+                    'url': target_url
+                })
+                record.share_link = tracker.short_url
+                record.link_tracker_id = tracker
+            else:
+                record.share_link = target_url
+
+    @api.depends('coupon_id', 'program_id')
+    def _get_related_links(self):
+        for record in self:
+            record.related_link_ids = self.env['link.tracker'].search([
+            ])
+            record.related_link_count = len(record.related_share_ids)
+
+    @api.onchange('related_link_ids')
+    def _set_related_links(self):
+        for record in self:
+            self.env['link.tracker'].search([
+                ('id', 'not in', record.related_link_ids.ids),
+            ]).unlink()
+
+    @api.onchange('website_id')
+    def _set_shop_url(self):
+        self.ensure_one()
+        self.redirect = '{base}/shop'.format(base=self.website_id.get_base_url()) if self.website_id else ''
+
+    @api.model
+    def create_share_action(self, coupon=None, program=None):
+        if bool(program) == bool(coupon):
+            raise UserError('Provide either a coupon or a program')
```

Comment:
Missing translation :( 
Finish your sentence by a dot `.`
Use double quote for string that will be displayed

```suggestion
            raise UserError(_("Provide either a coupon or a program."))
```